proxy_http.go

```diff
@@ -11,169 +12,196 @@ import (
-	discardFirstRequest bool
-	idleTimeout         time.Duration
-	onRequest           func(req *http.Request) *http.Request
-	onResponse          func(resp *http.Response) *http.Response
-	onError             func(req *http.Request, err error) *http.Response
-	dial                DialFunc
+func upstreamAddr(ctx context.Context) string {
+	upstreamAddr := ctx.Value(ctxKeyUpstreamAddr)
+	if upstreamAddr == nil {
+		return ""
+	}
+	return upstreamAddr.(string)
 }
 
-func (ic *httpInterceptor) intercept(w http.ResponseWriter, req *http.Request) error {
-	var downstream net.Conn
-	var downstreamBuffered *bufio.ReadWriter
-	tr := &http.Transport{
-		Dial: func(net, addr string) (net.Conn, error) {
-			return ic.dial(net, addr)
-		},
-		IdleConnTimeout: ic.idleTimeout,
-		// since we have one transport per downstream connection, we don't need
-		// more than this
-		MaxIdleConnsPerHost: 1,
+func (opts *Opts) applyHTTPDefaults() {
+	// Apply defaults
+	if opts.Filter == nil {
+		opts.Filter = filters.FilterFunc(defaultFilter)
 	}
-	var err error
+	if opts.OnError == nil {
+		opts.OnError = defaultOnError
+	}
+	if opts.IdleTimeout > 0 {
+		opts.Filter = filters.Join(filters.FilterFunc(func(ctx context.Context, req *http.Request, next filters.Next) (*http.Response, context.Context, error) {
+			resp, nextCtx, err := next(ctx, req)
+			if resp != nil {
+				opts.addIdleKeepAlive(resp.Header)
+			}
+			return resp, nextCtx, err
+		}), opts.Filter)
+	}
+}
 
-	closeDownstream := false
+// Handle implements the interface Proxy
+func (proxy *proxy) Handle(ctx context.Context, downstream net.Conn) error {
 	defer func() {
-		if closeDownstream {
-			if closeErr := downstream.Close(); closeErr != nil {
-				log.Tracef("Error closing downstream connection: %s", closeErr)
-			}
+		if closeErr := downstream.Close(); closeErr != nil {
+			log.Tracef("Error closing downstream connection: %s", closeErr)
 		}
-		tr.CloseIdleConnections()
 	}()
 
-	// Hijack underlying connection.
-	downstream, downstreamBuffered, err = w.(http.Hijacker).Hijack()
+	downstreamBuffered := bufio.NewReader(downstream)
+	ctx = context.WithValue(ctx, ctxKeyDownstream, downstream)
+	ctx = context.WithValue(ctx, ctxKeyRequestNumber, 1)
+
+	// Read initial request
+	req, err := http.ReadRequest(downstreamBuffered)
+	if req != nil {
+		remoteAddr := downstream.RemoteAddr()
+		if remoteAddr != nil {
+			req.RemoteAddr = downstream.RemoteAddr().String()
+		}
+	}
 	if err != nil {
-		fullErr := errors.New("Unable to hijack connection: %s", err)
-		respondBadGateway(w, fullErr)
-		return fullErr
+		errResp := proxy.OnError(ctx, req, true, err)
+		if errResp != nil {
+			proxy.writeResponse(downstream, req, errResp)
+		}
+		return err
+	}
+
+	var next filters.Next
+	if req.Method == http.MethodConnect {
+		next = proxy.nextCONNECT(downstream)
+	} else {
+		tr := &http.Transport{
+			DialContext: func(ctx context.Context, net, addr string) (net.Conn, error) {
+				return proxy.Dial(false, net, addr)
```

Comment:
Good idea. I'm actually just using whatever timeout the caller passed in, even if it's bigger than 30.